### ISSUE TITLE
Improve pack eval accuracy: 7 specific improvements to close the gap on losing packs

### ISSUE THREAD

**Issue by rysweet** [rysweet/agent-kgpacks#211]
# Improving Knowledge Pack Eval Accuracy

## Current State (46/48 packs evaluated, 2,314 questions)

| Metric | Value |
|--------|-------|
| Training accuracy (Claude Opus alone) | 91.4% |
| Pack accuracy (KG Agent with enhancements) | 96.7% |
| Delta | +5.3pp |
| Pack wins | 29 of 46 (63%) |
| Pack ties | 9 of 46 (20%) |
| **Pack losses** | **8 of 46 (17%)** |

The pack system adds significant value overall (+5.3pp), but **8 packs score WORSE than Claude alone**. This issue documents why that happens and proposes 7 specific fixes.

## The 8 Losing Packs

| Pack | Questions | Training | Pack | Delta | Root Cause |
|------|-----------|----------|------|-------|-----------|
| go-expert | 50 | 100% | 94% | -6pp | Claude already knows Go perfectly; pack injects irrelevant sections |
| llamaindex-expert | 50 | 98% | 94% | -4pp | Smallest pack DB (22MB); thin content from docs.llamaindex.ai |
| vercel-ai-sdk | 50 | 94% | 90% | -4pp | JS-heavy pages produce thin extracted content |
| zig-expert | 50 | 96% | 92% | -4pp | Zig docs are sparse; some URLs returned minimal content |
| bicep-infrastructure | 51 | 96% | 94% | -2pp | Content overlap with general Azure docs confuses retrieval |
| langchain-expert | 50 | 100% | 98% | -2pp | 20/36 URLs returned 404 (LangChain changed their URL structure) |
| openai-api-expert | 50 | 98% | 96% | -2pp | OpenAI blocks automated fetchers (HTTP 403); sparse content |
| react-expert | 50 | 100% | 98% | -2pp | Claude already knows React 19 very well from training |

### Pattern Analysis

The losses fall into two categories:

1. **Claude already knows it perfectly** (go, react, langchain, terraform): Training is 100%. The pack can only hurt by injecting irrelevant or misleading context. These are well-known, heavily-documented technologies that are thoroughly covered in Claude's training data.

2. **Pack content is thin or broken** (llamaindex, vercel-ai-sdk, zig, openai-api): The source URLs either returned errors (403, 404), produced thin content (JS-rendered pages), or the domain has sparse documentation. The vector search finds low-relevance matches that confuse synthesis.

### Problem

The synthesis prompt ALWAYS injects retrieved content, regardless of how relevant it is. When vector similarity is low (0.2-0.3), the retrieved sections are irrelevant noise. The "augment-not-replace" prompt says "use BOTH your expertise AND retrieved content" — but when the retrieved content is wrong, Claude may be misled by it.

### Evidence

For go-expert (where training is 100% and pack drops to 94%), the failure mode is: vector search returns vaguely related Go content (e.g., a section about Go modules when the question is about Go generics), Claude includes some wrong details from that section in its answer, and the judge scores it lower.

### Fix

In `wikigr/agent/kg_agent.py`, in the `query()` method, after vector search returns results:

```python
# After vector search
max_similarity = max(r.get("similarity", 0.0) for r in vector_results)

if max_similarity < self.CONTEXT_CONFIDENCE_THRESHOLD:
    # Low confidence: let Claude answer from training, mention pack was consulted
    logger.info(
        "Low retrieval confidence (%.2f < %.2f), deferring to training knowledge",
        max_similarity, self.CONTEXT_CONFIDENCE_THRESHOLD
    )
    # Skip pack context injection — use a minimal prompt
    synthesis_context = (
        "The knowledge pack was consulted but no highly relevant content was found "
        "for this specific question. Please answer using your own expertise."
    )
else:
    # High confidence: use normal retrieval-augmented synthesis
    synthesis_context = self._build_synthesis_context(question, kg_results, ...)
```

Add class constant:
```python
CONTEXT_CONFIDENCE_THRESHOLD = 0.5  # Skip pack context when similarity < 0.5
```

### Expected Impact

This would eliminate ALL negative deltas on packs where training is already strong (go, react, langchain). When the pack has nothing useful to add, it gets out of the way.

### Testing

1. Run go-expert eval with confidence gating at 0.5 — should see 100% match training
2. Run azure-ai-foundry eval — should still show +18pp (its similarities are high)
3. A/B test different thresholds: 0.4, 0.5, 0.6 — measure the optimal cutoff

### Files to Change

- `wikigr/agent/kg_agent.py`: Add `CONTEXT_CONFIDENCE_THRESHOLD`, modify `query()` method
- `tests/agent/test_kg_agent_core.py`: Add test for confidence gating behavior

---

### Problem

Several packs were built with test-mode (5 URLs) or had high URL failure rates:

| Pack | Current Articles | Total URLs | URL Failures | Issue |
|------|-----------------|------------|-------------|-------|
| llamaindex-expert | ~5 | 33 | Unknown | Smallest DB (22MB) |
| langchain-expert | ~16 | 36 | 20 (56%) | LangChain changed URL structure |
| openai-api-expert | ~5 | 35 | Many | OpenAI returns 403 for automated requests |
| vercel-ai-sdk | ~28 | 38 | 8 | JS-heavy pages produce thin content |
| zig-expert | ~5 | 38 | Unknown | Zig docs are sparse |

### Fix

For each losing pack:

1. **langchain-expert**: Update `data/packs/langchain-expert/urls.txt` with current LangChain documentation URLs (they moved from `docs.langchain.com` to `python.langchain.com/docs/`). Rebuild.

2. **openai-api-expert**: Use a browser-based fetcher (playwright/selenium) instead of `requests` to handle the 403 blocks. OR use the OpenAI cookbook/examples on GitHub which are publicly accessible.

3. **vercel-ai-sdk**: Use `sdk.vercel.ai/docs` URLs which are server-rendered. The current URLs may be hitting client-side-rendered pages.

4. **llamaindex-expert**: Expand the URL list to cover more of the LlamaIndex documentation (currently only 33 URLs, some failed).

5. **zig-expert**: Add URLs from `zig.guide` and community resources, not just `ziglang.org`.

### Expected Impact

More content = better vector search matches = higher relevance = less confusion in synthesis. The physics pack (451 articles, 100% accuracy) proves that content depth matters.

### Files to Change

- `data/packs/{pack}/urls.txt` for each losing pack
- Rebuild each pack: `echo "y" | uv run python scripts/build_{pack}_pack.py`

---

### Problem

The bi-encoder (BGE bge-base-en-v1.5) embeds queries and documents independently, then compares via cosine similarity. This misses nuanced relevance signals. For example, "How do you handle nullable types in Go?" might match a section about "Go type system" (semantically close) but miss a section titled "Handling nil values" (different wording but more relevant).

### Fix

After initial vector retrieval returns top-K results (currently K=10), run a cross-encoder to rerank them. Cross-encoders see BOTH the query and the document together, producing much more accurate relevance scores.

```python
from sentence_transformers import CrossEncoder

class CrossEncoderReranker:
    """Rerank vector search results using a cross-encoder for higher precision."""
    
    MODEL = "cross-encoder/ms-marco-MiniLM-L-12-v2"
    
    def __init__(self):
        self.model = CrossEncoder(self.MODEL)
    
    def rerank(self, query: str, results: list[dict], top_k: int = 5) -> list[dict]:
        """Rerank results by cross-encoder relevance score."""
        if not results:
            return results
        
        # Create query-document pairs
        pairs = [(query, r.get("content", r.get("title", ""))) for r in results]
        
        # Score all pairs
        scores = self.model.predict(pairs)
        
        # Sort by cross-encoder score
        scored = sorted(zip(scores, results), key=lambda x: x[0], reverse=True)
        return [r for _, r in scored[:top_k]]
```

### Integration Point

In `_vector_primary_retrieve()`, after `semantic_search()` returns results:

```python
vector_results = self.semantic_search(question, top_k=max_results * 2)  # Fetch more candidates
if self.cross_encoder:
    vector_results = self.cross_encoder.rerank(question, vector_results, top_k=max_results)
```

### Expected Impact

Research consistently shows cross-encoder reranking improves RAG retrieval precision by 10-15%. The cost is ~50ms per rerank call (negligible vs the 10-15 second Opus synthesis call). The `ms-marco-MiniLM-L-12-v2` model is only 33MB and runs on CPU.

### Testing

1. Run eval on the 8 losing packs with and without cross-encoder reranking
2. Measure retrieval precision: for each question, check if the top-1 result after reranking is more relevant than the top-1 from pure vector search
3. Measure latency impact: cross-encoder should add < 100ms per query

### Files to Change

- New file: `wikigr/agent/cross_encoder.py`
- `wikigr/agent/kg_agent.py`: Add optional cross-encoder to `__init__`, use in `_vector_primary_retrieve()`
- `tests/agent/test_cross_encoder.py`: Unit tests with mocked cross-encoder

---

### Problem

A single question embedding may miss relevant content phrased differently. "How do you handle errors in Go?" won't match a section titled "Panic and recover patterns" because the embeddings encode different vocabulary.

### Fix

Generate 2-3 reformulations of each question before searching, take the union of results:

```python
def _multi_query_retrieve(self, question: str, max_results: int = 5) -> list[dict]:
    """Search with multiple query reformulations for better recall."""
    queries = [question]
    
    # Generate reformulations via Claude (cached per question)
    try:
        response = self.claude.messages.create(
            model="claude-haiku-4-5-20251001",  # Use Haiku for speed/cost
            max_tokens=150,
            messages=[{"role": "user", "content": (
                "Generate 2 alternative phrasings of this question for document search. "
                "Return ONLY a JSON array of strings.\n\n"
                f"Question: {question}"
            )}]
        )
        alternatives = json.loads(response.content[0].text)
        queries.extend(alternatives[:2])
    except Exception:
        pass  # Fall back to single query
    
    # Search with all queries, deduplicate by title
    all_results = {}
    for q in queries:
        for r in self.semantic_search(q, top_k=max_results):
            title = r["title"]
            if title not in all_results or r["similarity"] > all_results[title]["similarity"]:
                all_results[title] = r
    
    return sorted(all_results.values(), key=lambda x: x["similarity"], reverse=True)[:max_results]
```

### Expected Impact

Multi-query retrieval improves recall by 15-25% in RAG benchmarks. The cost is 1 additional Haiku call (~$0.001) + 2 extra vector searches (~0ms). For the 8 losing packs where the single query misses the right content, this could recover the correct sections.

### Testing

1. Trace specific failing questions (e.g., the go-expert questions that score 0)
2. Check if the multi-query reformulations would have found the correct section
3. Measure recall@5 with single vs multi-query

### Files to Change

- `wikigr/agent/kg_agent.py`: Add `_multi_query_retrieve()`, use in `_vector_primary_retrieve()` as optional enhancement

---

### Problem

Not all retrieved sections are equally useful. A 20-word "See also" stub and a 500-word technical explanation both get passed to synthesis equally. The stub wastes context window and can dilute the useful content.

### Fix

Score each retrieved section before including it in the synthesis prompt:

```python
def _score_section_quality(self, content: str, question: str) -> float:
    """Score a section's quality for inclusion in synthesis context.
    
    Returns 0.0-1.0 where 1.0 is highest quality.
    """
    score = 0.0
    word_count = len(content.split())
    
    # Length score (penalize very short or very long sections)
    if word_count < 20:
        return 0.0  # Too short to be useful
    elif word_count < 50:
        score += 0.2
    elif word_count < 200:
        score += 0.5
    else:
        score += 0.8
    
    # Keyword overlap (does section mention query terms?)
    query_terms = set(question.lower().split()) - STOP_WORDS
    section_terms = set(content.lower().split())
    overlap = len(query_terms & section_terms) / max(len(query_terms), 1)
    score += overlap * 0.2
    
    return min(1.0, score)
```

Filter sections below a quality threshold before synthesis:

```python
scored_sections = [(self._score_section_quality(s, question), s) for s in sections]
good_sections = [s for score, s in scored_sections if score >= 0.3]
```

### Expected Impact

Filtering out low-quality sections prevents the "noise injection" problem that causes the losing packs to score below training. When only high-quality, relevant sections are injected, the synthesis stays on track.

### Files to Change

- `wikigr/agent/kg_agent.py`: Add `_score_section_quality()`, use in `_build_synthesis_context()`

---

### Problem

Specific URL issues per losing pack:

### langchain-expert
```
# 20 of 36 URLs return 404
# OLD: https://docs.langchain.com/docs/...
# NEW: https://python.langchain.com/docs/...
```
The LangChain project restructured their documentation. Every `docs.langchain.com` URL needs updating to `python.langchain.com`.

Action: Update `data/packs/langchain-expert/urls.txt` with new URLs.

### openai-api-expert
```
# Most platform.openai.com URLs return 403 for automated fetchers
# The pages require JavaScript rendering
```
Action: Replace `platform.openai.com` URLs with GitHub-hosted content:
- `https://github.com/openai/openai-cookbook/tree/main/examples/`
- `https://github.com/openai/openai-python` README
- `https://raw.githubusercontent.com/openai/openai-cookbook/main/...`

### vercel-ai-sdk
```
# sdk.vercel.ai uses heavy JavaScript rendering
# Many pages return minimal content when fetched with requests
```
Action: Target the non-JS documentation pages or use the GitHub source:
- `https://github.com/vercel/ai/tree/main/packages/ai/core`
- `https://github.com/vercel/ai/blob/main/README.md`

### Files to Change

- `data/packs/langchain-expert/urls.txt`
- `data/packs/openai-api-expert/urls.txt`
- `data/packs/vercel-ai-sdk/urls.txt`
- Rebuild each pack after URL fixes

---

### Problem

Some eval questions may be poorly calibrated:
- Expected answers may use terminology that differs from pack content
- Questions may test knowledge not covered by the pack's URL scope
- "Hard" questions may require multi-hop reasoning the pack doesn't support

### Fix

For each losing pack, review the questions that scored 0 or below 7:

```bash
# Find low-scoring questions
python3 -c "
import json
d = json.load(open('/tmp/eval_full_final/go-expert.json'))
for i, (ts, ps) in enumerate(zip(d['training']['scores'], d['pack']['scores'])):
    if ps < ts:
        print(f'Q{i}: training={ts}, pack={ps}')
"
```

Then for each low-scoring question:
1. Check if the pack actually contains content that could answer it
2. Check if the expected answer matches the pack's content terminology
3. If the question is out of scope for the pack, mark it as such

### Expected Impact

Removing or fixing miscalibrated eval questions gives a more accurate picture of actual pack quality. Some "losses" may be eval artifacts rather than real quality issues.

### Files to Change

- `data/packs/{pack}/eval/questions.jsonl` for each losing pack
- May need to regenerate questions with better domain descriptions

---

## Implementation Priority

| # | Improvement | Effort | Expected Impact | Dependencies |
|---|------------|--------|----------------|-------------|
| 1 | Confidence-gated context injection | Low (20 LOC) | Eliminates ALL negative deltas on strong-training packs | None |
| 2 | Full pack rebuilds for thin packs | Medium (URL research + rebuild) | +4-6pp on thin packs | None |
| 3 | Cross-encoder reranking | Medium (new module) | +10-15% retrieval precision | None |
| 4 | Multi-query retrieval | Low (30 LOC) | +15-25% recall | None |
| 5 | Content quality scoring | Low (20 LOC) | Prevents noise injection | None |
| 6 | Fix broken URL lists | Low (URL research) | Fixes langchain, openai, vercel | Rebuild after |
| 7 | Eval question quality review | Medium (manual) | More accurate measurement | None |

**Recommended order**: 1 → 6 → 2 → 5 → 3 → 4 → 7

Improvement 1 alone would likely bring the 8 losing packs to parity with training, pushing overall accuracy from 96.7% to ~98%+.

---

## Current Eval Data (46/48 packs, 2,314 questions)

Dotnet-expert (202q) and rust-expert (200q) are still running. Full results for 46 packs available in `/tmp/eval_full_final/`.
